upstairs/src/volume.rs

```diff
@@ -684,100 +678,35 @@ impl BlockIO for Volume {
                     .as_ref()
                     .unwrap()
-                    .read(parent_offset, parent_buffer.clone())
+                    .read(parent_offset, Buffer::new(sub_sz))
                     .await?;
 
-                let mut data_vec = data.as_vec().await;
-                let mut owned_vec = data.owned_vec().await;
+                data.eat(data_index, parent_buffer);
+            }
 
-                let sub_data_vec = sub_buffer.as_vec().await;
-                let sub_owned_vec = sub_buffer.owned_vec().await;
-                let parent_data_vec = parent_buffer.as_vec().await;
+            let sub_offset = Block::new(
+                sub_volume.compute_sub_volume_lba(coverage.start),
+                offset.shift,
+            );
+            let sz = (coverage.end - coverage.start) as usize
+                * self.block_size as usize;
 
-                // "ownership" comes back from the downstairs per byte but all
-                // writes occur per block. Iterate over the blocks that both the
-                // read-only parent and subvolume cover here.
-                //
-                // This layer of the volume "owns" a block if the subvolume
-                // "owns" the block, *not* if the read only parent does.
-                for i in 0..(sub_sz as u64 / self.block_size) {
-                    let start_of_block = (i * self.block_size) as usize;
-
-                    if sub_owned_vec[start_of_block] {
-                        // sub volume has written to this block, use it
-                        for block_offset in 0..self.block_size {
-                            let inside_block =
-                                start_of_block + block_offset as usize;
-
-                            data_vec[data_index + inside_block] =
-                                sub_data_vec[inside_block];
-
-                            // this layer of the volume does own this
-                            // block, it was written to the subvolume
-                            owned_vec[data_index + inside_block] = true;
-                        }
-                    } else {
-                        // sub volume hasn't written to this block, use the
-                        // parent
-                        for block_offset in 0..self.block_size {
-                            let inside_block =
-                                start_of_block + block_offset as usize;
-
-                            data_vec[data_index + inside_block] =
-                                parent_data_vec[inside_block];
-
-                            // this layer of the volume doesn't own this
-                            // block, it came from the read only parent. Note
-                            // this doesn't depend if the block was owned by
-                            // the read only parent, just that the subvolume
-                            // doesn't own it.
-                            owned_vec[data_index + inside_block] = false;
```

Comment:
I believe this behavior has changed: because we're using `Buffer::eat` to load the parent buffer (at L688), the new implementation will set `owned = true` for bytes coming from the parent.

Reply:
You're not wrong, but I believe (after sleeping on it for a few nights) that the original behaviour might have been a bug! Ownership should come from the downstairs: has this block been written to or not? This referenced line *does not* take ownership information from `parent_buffer`, it overwrites it to false.

This test fails on main:

```patch
diff --git a/upstairs/src/volume.rs b/upstairs/src/volume.rs
index e388767..bf21fac 100644
--- a/upstairs/src/volume.rs
+++ b/upstairs/src/volume.rs
@@ -2150,6 +2150,76 @@ mod test {
         Ok(())
     }
 
+    #[tokio::test]
+    async fn is_it_a_bug_or_not() -> Result<()> {
+        const BLOCK_SIZE: u64 = 512;
+
+        let parent = Arc::new(InMemoryBlockIO::new(
+            Uuid::new_v4(),
+            BLOCK_SIZE,
+            2048,
+        ));
+
+        // Ownership starts off false
+
+        let buffer = Buffer::new(1024);
+        parent
+            .read(Block::new(0, BLOCK_SIZE.trailing_zeros()), buffer.clone())
+            .await?;
+
+        assert_eq!(&*buffer.owned_vec().await, &[false; 1024]);
+
+        // Ownership is set by writing to blocks
+
+        parent
+            .write(
+                Block::new(0, BLOCK_SIZE.trailing_zeros()),
+                Bytes::from(vec![9; 512]),
+            )
+            .await?;
+
+        // Ownership is returned by the downstairs
+
+        let buffer = Buffer::new(1024);
+        parent
+            .read(Block::new(0, BLOCK_SIZE.trailing_zeros()), buffer.clone())
+            .await?;
+
+        let mut expected = vec![9u8; 512];
+        expected.extend(vec![0u8; 512]);
+
+        assert_eq!(&*buffer.as_vec().await, &expected);
+
+        let mut expected_ownership = vec![true; 512];
+        expected_ownership.extend(vec![false; 512]);
+
+        assert_eq!(&*buffer.owned_vec().await, &expected_ownership);
+
+        // Ownership through a volume should be the same!!
+
+        let disk = Arc::new(InMemoryBlockIO::new(
+            Uuid::new_v4(),
+            BLOCK_SIZE,
+            4096,
+        ));
+
+        let mut volume = Volume::new(BLOCK_SIZE, csl());
+        volume.add_subvolume(disk).await?;
+        volume.add_read_only_parent(parent.clone()).await?;
+
+        // So is it?
+
+        let buffer = Buffer::new(1024);
+        volume
+            .read(Block::new(0, BLOCK_SIZE.trailing_zeros()), buffer.clone())
+            .await?;
+
+        assert_eq!(&*buffer.as_vec().await, &expected);
+        assert_eq!(&*buffer.owned_vec().await, &expected_ownership);
+
+        Ok(())
+    }
+
     #[tokio::test]
     async fn test_parent_uninitialized_read_only_region_one_subvolume(
     ) -> Result<()> {
```

This passes on this branch:

```patch
diff --git a/upstairs/src/lib.rs b/upstairs/src/lib.rs
index d88b8ff..848cc24 100644
--- a/upstairs/src/lib.rs
+++ b/upstairs/src/lib.rs
@@ -1568,6 +1568,10 @@ impl Buffer {
             }
         }
     }
+
+    pub fn owned_ref(&self) -> &[bool] {
+        &self.owned
+    }
 }
 
 impl std::ops::Deref for Buffer {
diff --git a/upstairs/src/volume.rs b/upstairs/src/volume.rs
index 4051f40..4a60ca9 100644
--- a/upstairs/src/volume.rs
+++ b/upstairs/src/volume.rs
@@ -2077,6 +2077,76 @@ mod test {
         Ok(())
     }
 
+    #[tokio::test]
+    async fn is_it_a_bug_or_not() -> Result<()> {
+        const BLOCK_SIZE: u64 = 512;
+
+        let parent = Arc::new(InMemoryBlockIO::new(
+            Uuid::new_v4(),
+            BLOCK_SIZE,
+            2048,
+        ));
+
+        // Ownership starts off false
+
+        let buffer = Buffer::new(1024);
+        let buffer = parent
+            .read(Block::new(0, BLOCK_SIZE.trailing_zeros()), buffer)
+            .await?;
+
+        assert_eq!(buffer.owned_ref(), &[false; 1024]);
+
+        // Ownership is set by writing to blocks
+
+        parent
+            .write(
+                Block::new(0, BLOCK_SIZE.trailing_zeros()),
+                Bytes::from(vec![9; 512]),
+            )
+            .await?;
+
+        // Ownership is returned by the downstairs
+
+        let buffer = Buffer::new(1024);
+        let buffer = parent
+            .read(Block::new(0, BLOCK_SIZE.trailing_zeros()), buffer)
+            .await?;
+
+        let mut expected = vec![9u8; 512];
+        expected.extend(vec![0u8; 512]);
+
+        assert_eq!(&*buffer, &expected);
+
+        let mut expected_ownership = vec![true; 512];
+        expected_ownership.extend(vec![false; 512]);
+
+        assert_eq!(buffer.owned_ref(), &expected_ownership);
+
+        // Ownership through a volume should be the same!!
+
+        let disk = Arc::new(InMemoryBlockIO::new(
+            Uuid::new_v4(),
+            BLOCK_SIZE,
+            4096,
+        ));
+
+        let mut volume = Volume::new(BLOCK_SIZE, csl());
+        volume.add_subvolume(disk).await?;
+        volume.add_read_only_parent(parent.clone()).await?;
+
+        // So is it?
+
+        let buffer = Buffer::new(1024);
+        let buffer = volume
+            .read(Block::new(0, BLOCK_SIZE.trailing_zeros()), buffer)
+            .await?;
+
+        assert_eq!(&*buffer, &expected);
+        assert_eq!(&*buffer.owned_ref(), &expected_ownership);
+
+        Ok(())
+    }
+
     #[tokio::test]
     async fn test_parent_uninitialized_read_only_region_one_subvolume(
     ) -> Result<()> {
```

Reply:
Okay, I agree that the new behavior seems correct!

Reply:
I've added this as a test in 9bd6061